clash-cores/test/Test/Cores/Xilinx/DcFifo.hs

```diff
@@ -65,8 +76,84 @@ tests = testGroup "FIFO tests"
+    , testCase "test_feedback_F2_F5" (test_feedback @F2 @F5 Proxy Proxy)
+    , testCase "test_feedback_F2_F6" (test_feedback @F2 @F6 Proxy Proxy)
+    , testCase "test_feedback_F3_F1" (test_feedback @F3 @F1 Proxy Proxy)
+    , testCase "test_feedback_F3_F2" (test_feedback @F3 @F2 Proxy Proxy)
+    , testCase "test_feedback_F3_F3" (test_feedback @F3 @F3 Proxy Proxy)
+    , testCase "test_feedback_F3_F4" (test_feedback @F3 @F4 Proxy Proxy)
+    , testCase "test_feedback_F3_F5" (test_feedback @F3 @F5 Proxy Proxy)
+    , testCase "test_feedback_F3_F6" (test_feedback @F3 @F6 Proxy Proxy)
+    , testCase "test_feedback_F4_F1" (test_feedback @F4 @F1 Proxy Proxy)
+    , testCase "test_feedback_F4_F2" (test_feedback @F4 @F2 Proxy Proxy)
+    , testCase "test_feedback_F4_F3" (test_feedback @F4 @F3 Proxy Proxy)
+    , testCase "test_feedback_F4_F4" (test_feedback @F4 @F4 Proxy Proxy)
+    , testCase "test_feedback_F4_F5" (test_feedback @F4 @F5 Proxy Proxy)
+    , testCase "test_feedback_F4_F6" (test_feedback @F4 @F6 Proxy Proxy)
+    , testCase "test_feedback_F5_F1" (test_feedback @F5 @F1 Proxy Proxy)
+    , testCase "test_feedback_F5_F2" (test_feedback @F5 @F2 Proxy Proxy)
+    , testCase "test_feedback_F5_F3" (test_feedback @F5 @F3 Proxy Proxy)
+    , testCase "test_feedback_F5_F4" (test_feedback @F5 @F4 Proxy Proxy)
+    , testCase "test_feedback_F5_F5" (test_feedback @F5 @F5 Proxy Proxy)
+    , testCase "test_feedback_F5_F6" (test_feedback @F5 @F6 Proxy Proxy)
+    , testCase "test_feedback_F6_F1" (test_feedback @F6 @F1 Proxy Proxy)
+    , testCase "test_feedback_F6_F2" (test_feedback @F6 @F2 Proxy Proxy)
+    , testCase "test_feedback_F6_F3" (test_feedback @F6 @F3 Proxy Proxy)
+    , testCase "test_feedback_F6_F4" (test_feedback @F6 @F4 Proxy Proxy)
+    , testCase "test_feedback_F6_F5" (test_feedback @F6 @F5 Proxy Proxy)
+    , testCase "test_feedback_F6_F6" (test_feedback @F6 @F6 Proxy Proxy)
+    ]
   ]
 
+-- | Tests whether 'dcFifo's over and underflow signals can be used in a
+-- feedback loop with 'dualFlipFlopSynchronizer'.
+test_feedback ::
+  forall readDom writeDom .
+  (KnownDomain readDom, KnownDomain writeDom) =>
+  Proxy readDom ->
+  Proxy writeDom ->
+  Assertion
+test_feedback Proxy Proxy = do
+  () <- pure $ rnf $ P.last $ sampleN 1000 overflowSynced
+  () <- pure $ rnf $ P.last $ sampleN 1000 underflowSynced
+  pure ()
+ where
+  (clkR, rstR, enaR) = (clockGen @readDom, resetGen, enableGen)
+  (clkW, rstW, enaW) = (clockGen @writeDom, resetGen, enableGen)
+
+  writes = pure (Just True)
+  noWrites = pure Nothing
+  writeMaybe = bool <$> noWrites <*> writes <*> writeEnable
+
+  config = (defConfig @8){dcOverflow=True, dcUnderflow=True}
+  FifoOut{..} = dcFifo config clkW rstW clkR rstR writeMaybe readEnable
+  overflowSynced = dualFlipFlopSynchronizer clkW clkR rstR enaR False isOverflow
+  underflowSynced = dualFlipFlopSynchronizer clkR clkW rstW enaW False isUnderflow
+
+  readEnable =
+    delay clkR enaR True $
```

Comment:
I think you can remove the `delay` here and on line 154, makes the whole thing a little simpler!